src/inference/hmc.jl

```diff
@@ -48,190 +48,394 @@ mutable struct HMC{AD, T} <: StaticHamiltonian{AD}
     epsilon   ::  Float64   # leapfrog step size
     tau       ::  Int       # leapfrog step number
     space     ::  Set{T}    # sampling space, emtpy means all
+    metricT
 end
 HMC(args...) = HMC{ADBackend()}(args...)
-function HMC{AD}(epsilon::Float64, tau::Int, space...) where AD
+function HMC{AD}(n_iters::Int, epsilon::Float64, tau::Int; metricT=AHMC.UnitEuclideanMetric) where AD
+    return HMC{AD, Any}(n_iters, epsilon, tau, Set(), metricT)
+end
+function HMC{AD}(n_iters::Int, epsilon::Float64, tau::Int, space...; metricT=AHMC.UnitEuclideanMetric) where AD
     _space = isa(space, Symbol) ? Set([space]) : Set(space)
-    return HMC{AD, eltype(_space)}(1, epsilon, tau, _space)
+    return HMC{AD, eltype(_space)}(n_iters, epsilon, tau, _space, metricT)
 end
-function HMC{AD}(n_iters::Int, epsilon::Float64, tau::Int) where AD
-    return HMC{AD, Any}(n_iters, epsilon, tau, Set())
+
+"""
+    HMCDA(n_iters::Int, n_adapts::Int, δ::Float64, λ::Float64; init_ϵ::Float64=0.0)
+
+Hamiltonian Monte Carlo sampler with Dual Averaging algorithm.
+
+Usage:
+
+```julia
+HMCDA(1000, 200, 0.65, 0.3)
+```
+
+Arguments:
+
+- `n_iters::Int` : Number of samples to pull.
+- `n_adapts::Int` : Numbers of samples to use for adaptation.
+- `δ::Float64` : Target acceptance rate. 65% is often recommended.
+- `λ::Float64` : Target leapfrop length.
+- `init_ϵ::Float64=0.0` : Inital step size; 0 means automatically search by Turing.
+
+Example:
```

Comment:
Consider removing the following doc section since they are generic for all samplers. Perhaps we can transfer them to the `sample` API once https://github.com/TuringLang/Turing.jl/issues/746#issuecomment-485222516 is fixed?

```julia

Example:
```julia
# Define a simple Normal model with unknown mean and variance.
@model gdemo(x) = begin
  s ~ InverseGamma(2,3)
  m ~ Normal(0, sqrt(s))
  x[1] ~ Normal(m, sqrt(s))
  x[2] ~ Normal(m, sqrt(s))
  return s, m
end
sample(gdemo([1.j_max, 2]), NUTS(1000, 200, 0.6j_max))
```

Reply:
Good point. I was also thinking about removing the duplicated comments here. We should only leave examples for constructing algorithms here.